tests/UnifiedSpecTests/Constraint/Match.php

```diff
@@ -0,0 +1,470 @@
+<?php
+
+namespace MongoDB\Tests\UnifiedSpecTests\Constraint;
+
+use LogicException;
+use MongoDB\BSON\Type;
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+use MongoDB\Tests\UnifiedSpecTests\EntityMap;
+use PHPUnit\Framework\Constraint\Constraint;
+use PHPUnit\Framework\Constraint\LogicalOr;
+use RuntimeException;
+use SebastianBergmann\Comparator\ComparisonFailure;
+use SebastianBergmann\Comparator\Factory;
+use Symfony\Bridge\PhpUnit\ConstraintTrait;
+use function array_values;
+use function count;
+use function get_class;
+use function get_resource_type;
+use function gettype;
+use function hex2bin;
+use function implode;
+use function is_array;
+use function is_bool;
+use function is_object;
+use function is_resource;
+use function is_string;
+use function sprintf;
+use function stream_get_contents;
+use function strpos;
+use const PHP_INT_SIZE;
+
+/**
+ * Constraint that checks if one document matches another.
+ *
+ * The expected value is passed in the constructor.
+ */
+class Match extends Constraint
+{
+    use ConstraintTrait;
+
+    /** @var EntityMap */
+    private $entityMap;
+
+    /** @var boolean */
+    private $ignoreExtraKeysInRoot = false;
+
+    /** @var boolean */
+    private $ignoreExtraKeysInEmbedded = false;
+
+    /** @var mixed */
+    private $value;
+
+    /** @var ComparisonFailure|null */
+    private $lastFailure;
+
+    public function __construct($value, bool $ignoreExtraKeysInRoot = false, bool $ignoreExtraKeysInEmbedded = false, EntityMap $entityMap = null)
+    {
+        $this->value = self::prepare($value, true);
+        $this->entityMap = $entityMap ?? new EntityMap();
+        $this->ignoreExtraKeysInRoot = $ignoreExtraKeysInRoot;
+        $this->ignoreExtraKeysInEmbedded = $ignoreExtraKeysInEmbedded;
+        $this->comparatorFactory = Factory::getInstance();
+    }
+
+    public function evaluate($other, $description = '', $returnResult = false)
+    {
+        $other = self::prepare($other, true);
+        $success = false;
+        $this->lastFailure = null;
+
+        try {
+            $this->assertMatches($this->value, $other, $this->ignoreExtraKeysInRoot);
+            $success = true;
+        } catch (RuntimeException $e) {
+            $this->lastFailure = new ComparisonFailure(
+                $this->value,
+                $other,
+                $this->exporter()->export($this->value),
+                $this->exporter()->export($other),
+                false,
+                $e->getMessage()
+            );
+        }
+
+        if ($returnResult) {
+            return $success;
+        }
+
+        if (! $success) {
+            $this->fail($other, $description, $this->lastFailure);
+        }
+    }
+
+    private function assertEquals($expected, $actual, string $keyPath)
+    {
+        $expectedType = is_object($expected) ? get_class($expected) : gettype($expected);
```

Comment:
FYI, PHP 8 introduces `get_debug_type` to avoid these checks. If you want, we can introduce the [symfony/polyfill](https://github.com/symfony/polyfill) package to provide this for us.

Reply:
Great idea. Opened [PHPLIB-588](https://jira.mongodb.org/browse/PHPLIB-588).